python/ray/data/read_api.py

```diff
@@ -1073,6 +1074,50 @@ def convert(ds: "datasets.Dataset") -> Dataset[ArrowRow]:
         )
 
 
+@PublicAPI
+def read_image_folder(root: str, *, parallelism: int = -1):
```

Comment:
No strong opinion. 

If there's no consensus, I'll add the datasource now. We can always add a `read_*` API later.

Reply:
Let's do that.

Reply:
It might be worth adding this base API for the sake of discoverability.

Reply:
Relevant conversation on Slack:

**Balaji**
Hey, do you guys have any strong opinions about whether ImageFolderDatasource should have a corresponding read_* API?

**Clark**
I think it might be worth having, let’s discuss on the PR thread though: https://github.com/ray-project/ray/pull/24641/files#r921790567

**ekl**
I thought the issue was we were punting proper partitioning suppotr, so it doesn't make sense to have it first-class for now?

**Balaji**
By proper partitioning support, do you mean supporting arbitrary partitions, rather than assuming we partition class by the first directory after the root?

**ekl**
that's right (following standard pyarrow partitioning conventions)

**Balaji**
I think we'll be able to add proper partitioning support without making breaking changes to the read_* API

Since the functionality is useful without full partitioning support, maybe we should add a top-level API for the sake of discoverability (as Clark suggested on the thread)? (edited) 

**ekl**
I'm ok with that.

Reply:
Please, let's go with the consensus that was agreed on slack, just to avoid frustrations for others in the discussion.

Opening a new PR after with the read_* API shouldn't be too hard :)